File: 05_functions/python/closures.py
if __name__ == "__main__":
    example_closure = lambda x: x

    s = example_closure("hello")
    n = example_closure(5)
    print(s, n)

    x = [1, 2, 3]

    equal_to_x = lambda z: z == x

    print("this is x", x)
    y = [1, 2, 3]
    assert equal_to_x(y)

    items = [1, 2, 3, 4, 5]
    plus_one = list(map(lambda x: x + 1, items))
    sum_all = sum(map(lambda x: x + 1, items))
    print(plus_one, sum_all)

    two_args = lambda x, y: x - y
    print(two_args(5, 3))

File: 05_functions/rust/closures.rs
fn main() {
    let example_closure = |x| x;

    let s = example_closure(String::from("hello"));
    let n = example_closure(5);
    println!("{} {}", s, n);

    let x = vec![1, 2, 3];

    let equal_to_x = move |z| z == x;

    println!("this is x {:?}", x);
    let y = vec![1, 2, 3];
    assert!(equal_to_x(y));

    let items: Vec<i32> = vec![1, 2, 3, 4, 5];
    let plus_one: Vec<_> = items.iter().map(|x| x + 1).collect();
    let sum_all: i32 = items.iter().map(|x| x + 1).sum();
    println!("{:?} {}", plus_one, sum_all);

    let two_args = |x, y| x - y;
    println!("{}", two_args(5, 3));
}
